src/ophyd_async/core/_signal.py

```diff
@@ -582,37 +652,36 @@ async def set_and_wait_for_other_value(
         set_and_wait_for_value(device.acquire, 1, device.acquire_rbv, 1)
     """
     # Start monitoring before the set to avoid a race condition
-    values_gen = observe_value(read_signal)
+    values_gen = observe_value(match_signal)
 
     # Get the initial value from the monitor to make sure we've created it
     current_value = await anext(values_gen)
 
-    status = set_signal.set(set_value, timeout=set_timeout)
+    set_signal.set(set_value, timeout=set_timeout)
```

Comment:
@DominicOram this is what @ZohebShaikh was alluding to in https://github.com/bluesky/ophyd-async/issues/402#issuecomment-2357824352.

If we choose A, then we need to decide what to do with the Status. Something needs to keep track of it an await it, we can't drop it on the floor as we are doing here or we get teardown errors in the tests, so we have to return it. The calling code then looks like:
```python
status = await set_and_wait_for_other_value(dev.acquire, 1, dev.acquire_rbv, 1)
# do something that you can do as soon as the thing is acquiring
await status
# now the device has finished acquiring
```

If we choose B, then we keep a track of `status` here, then await it at the end of the function. The calling code becomes:
```python
await set_and_wait_for_other_value(dev.acquire, 1, dev.acquire_rbv, 1)
# now the device has finished acquiring,
```

What was your actual use case for this? Will the write_signal ever take significantly longer to caput-callback than the read_signal will take to change to the match_value? I know we have this case for the areaDetector acquire PV above, but we decided to not use this function as it was clearer to write:
```python
arm_status = dev.acquire.set(1)
await wait_for_value(dev.acquire_rbv, 1)
# do something we can do when the device is armed
await arm_status
# now the device is disarmed
```

Reply:
My original usecase for this was the actually the areaDetector one. We tried the code as suggested:
```python
arm_status = dev.acquire.set(1)
await wait_for_value(dev.acquire_rbv, 1)
await arm_status
```
But found that the `wait_for_value` was failing. This is because for this particular configuration the detector immediately takes data and does so so quickly that it has finished by the time we start monitoring the RBV. See https://github.com/bluesky/ophyd-async/issues/453#issue-2405971992. My main motivation for putting this into `ophyd-async` was to provide a function that would always protect against the potential race condition so we're less like to see it with people taking the naive approach.

I believe for my use case the callback has already returned on the set before the RBV goes to the expected value but I would have to check. Either way I think what I would expect to happen is that `set_and_wait_for_other_value` returns a gather of both statuses. If you want to do some in between these then you need to do that yourself. Maybe we should have a chat about it?